xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/CoinmateStreamingMarketDataService.java

```diff
@@ -51,22 +51,20 @@
 
   @Override
   public Observable<Trade> getTrades(CurrencyPair currencyPair, Object... args) {
-    String channelName = "trades-" + CoinmateStreamingAdapter.getChannelPostfix(currencyPair);
+    String channelName = "channel/trades/" + CoinmateStreamingAdapter.getChannelPostfix(currencyPair);
+
+    ObjectReader reader = StreamingObjectMapperHelper.getObjectMapper().readerFor(new TypeReference<List<CoinmateWebSocketTrade>>() {});
+    CoinmateStreamingService service = serviceFactory.createAndConnect(channelName, false);
 
-    return service
-        .subscribeChannel(channelName, "new_trades")
+    return service.subscribeMessages()
         .map(
             s -> {
-              ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
-              List<CoinmateWebSocketTrade> list =
-                  mapper.readValue(s, new TypeReference<List<CoinmateWebSocketTrade>>() {});
+              List<CoinmateWebSocketTrade> list = reader.readValue(s);
               return list;
             })
         .flatMapIterable(coinmateWebSocketTrades -> coinmateWebSocketTrades)
         .map(
-            coinmateWebSocketTrade ->
-                CoinmateAdapters.adaptTrade(
-                    coinmateWebSocketTrade.toTransactionEntry(
-                        CoinmateUtils.getPair(currencyPair))));
+            coinmateWebSocketTrade -> CoinmateStreamingAdapter.adaptTrade(coinmateWebSocketTrade, currencyPair)
```

Comment:
A few problems here:

- Since the socket is scoped to a single subscription, there needs to be an `onDispose` in all these methods to close the `StreamingService` when the `Observable` is disposed
- **However**, it's not a great idea to scope a socket to a single subscription.  The pattern used elsewhere is that the `Observable`s are `share()`ed, so if multiple subscriptions are made, the separate subscribers share the same socket. What I think you need to do here is maintain a `ConcurrentMap` of `CoinmateStreamingService`, re-use the existing one if it already exists and use `Observable.using()` to close the socket when the last subscriber disposes.

Reply:
Thanks for the suggestions, I am not that good with RxJava, so I am struggling to implement this. I made something like this, to handle both the locking and reusing of connections, but it only works for the first subscriber. I was thinking maybe a `ReplaySubject` would disposing work with it?

```
private ConcurrentMap<String, Observable<CoinmateStreamingService>> serviceMap;
  private static Lock connectionLock = new ReentrantLock();
  private Scheduler scheduler = Schedulers.single();
  public Observable<CoinmateStreamingService> createConnection(String endpoint, boolean needsAuth) {
    String url = baseUrl + "/" + endpoint;
    CoinmateStreamingService service;

    if (serviceMap.containsKey(url)) {
      return serviceMap.get(url);
    } else {
      String authUrl = url;
      // append auth parameters if necessary
      if (needsAuth && authParams != null) {
        authUrl += "?" + authParams.toParams();
      }
      service = new CoinmateStreamingService(authUrl);

      Observable<CoinmateStreamingService> observable = Observable.using(
          () -> service,
          s -> Completable.fromAction(() -> connectionLock.lock())
              .subscribeOn(scheduler)
              .andThen(s.connect())
              .observeOn(scheduler)
              .doFinally(() -> {
                connectionLock.unlock();
              }).andThen(Observable.just(s)),
          s -> s.disconnect()
      ).share();

      serviceMap.put(url, observable);

      return observable;
    }
  }
```

Reply:
It's not a case for `ReplaySubject`.  You're almost there. I;ll try and have a look at this when I get some time.

Reply:
Eventually I used a slightly different approach